stdlib/asm/math/u64.masm

```diff
@@ -673,6 +676,53 @@ export.unchecked_shr
     cswap
 end
 
+# Performs left shift of one unsigned 64-bit integer preserving the overflow and
+# using the pow2 operation.
+# The input value to be shifted is assumed to be represented using 32 bit limbs.
+# The shift value is assumed to be in the range [0, 64).
+# Stack transition looks as follows:
+# [b, a_hi, a_lo, ...] -> [d_hi, d_lo, c_hi, c_lo, ...], where (d,c) = a << b, 
+# which d contains the bits shifted out.
+# This takes 20 cycles.
+export.overflowing_shl
+    pow2
+    u32split
+    exec.overflowing_mul
+end
+
+# Performs right shift of one unsigned 64-bit integer preserving the overflow and
+# using the pow2 operation.
+# The input value to be shifted is assumed to be represented using 32 bit limbs.
+# The shift value is assumed to be in the range [0, 64).
+# Stack transition looks as follows:
+# [b, a_hi, a_lo, ...] -> [d_hi, d_lo, c_hi, c_lo, ...], where c = a >> b, d = a << (64 - b).
+# This takes 63 cycles.
+export.overflowing_shr
+    push.64             # (64 - b)
+    dup.1               
+    u32sub.unsafe
+    drop
+    movdn.3             # move result keeping [b, a_hi, a_lo, ....]
+    
+    dup.2               # dup a_lo, a_hi and bring b
+    dup.2
+    dup.2
+    exec.unchecked_shr  # c = a >> b
+
+    movup.4             # bring duplicated a_lo and a_hi
+    movup.4
+    movup.5             # bring the (64 - b) and b
+    movup.5
+    push.0.0.0.0        # adding zeroes on 0, 1, 2 and 3 to be able to use cdropw
```

Comment:
you can replace this with `padw`. won't save any cycles but would be a bit more concise.